### PR TITLE
[fix/RPA-1516.2] Fix credential collection resource

### PR DESCRIPTION
**Pull request by marcin-depcik**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 1]
```diff
@@ -81,56 +82,82 @@ export class CredentialCollectionService {
...
+                'filteredCredentialCollectionUsers.privilegeType = :privilegeType',
+                { privilegeType: PrivilegeType.WRITE },
+            )
+            .innerJoinAndSelect(
+                'credentialCollectionEntity.credentialCollectionUser',
+                'allCredentialCollectionUsers',
+            )
+            .innerJoinAndSelect('allCredentialCollectionUsers.user', 'allUsers')
+            .getMany();
     }
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: well congrats for achieving that in one query

but it's a huge change in the readability
1. You can merge these 3 'where' conditions into one - here example of using 2 conditions in one 'where' https://typeorm.io/select-query-builder#getting-values-using-querybuilder
2. I'm quite afraid for seeing so many inner joins in method that is going to be called very often for checking if user has access to some credential. But let's put off this problem for later when it'll arise.

Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: Do you really need to check createdBy? I think you're automatically add credential owner to credentialCollectionUser relation

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 2]
```diff
@@ -81,56 +82,82 @@ export class CredentialCollectionService {
...
+                { tenantId: user.tenantId },
+            )
+            .andWhere(
+                'filteredCredentialCollectionUsers.user.id = :userId',
+                { userId: user.id },
+            )
+            .andWhere(
+                'filteredCredentialCollectionUsers.privilegeType = :privilegeType',
+                { privilegeType: PrivilegeType.WRITE },
+            )
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: try to combine many 'where' into one

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 3]
```diff
@@ -213,10 +250,36 @@ export class CredentialCollectionService {
...
+            .findOne({
+                relations,
+                where: {
+                    id,
+                    tenantId: user.tenantId,
+                    credentialCollectionUser: {
+                        userId: user.id,
+                        privilegeType: PrivilegeType.WRITE,
+                    },
+                },
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts:251]: great you've added that check

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 4]
```diff
@@ -213,10 +250,36 @@ export class CredentialCollectionService {
...
+            });
+
+        if (!collection) {
+            throw new NotFoundException('Could not find credential collection with id: ' + id);
+        }
+
+        if (collection.credentials.length > 0) {
+            throw new ConflictException('Collection cannot be deleted, there are credentials related to this collection');
+        }
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts:260]: nice

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts (hunk 1 above)
Comment (marcin-depcik) [reply runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: I do not check `createdBy`, I retrieve relationships with this property

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 5]
```diff
@@ -100,60 +100,48 @@ export class CredentialCollectionService {
...
-            )
-            .andWhere(
-                'filteredCredentialCollectionUsers.user.id = :userId',
-                { userId: user.id },
-            )
-            .andWhere(
-                'filteredCredentialCollectionUsers.privilegeType = :privilegeType',
-                { privilegeType: PrivilegeType.WRITE },
-            )
+            .innerJoinAndSelect('credentialCollectionEntity.createdBy', 'credentialCollectionEntity_createdBy')
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: what  a monster: 'credentialCollectionEntity_createdBy'

Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: mixing snake case with dot case looks kinda strange

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 6]
```diff
@@ -100,60 +100,48 @@ export class CredentialCollectionService {
...
             .innerJoinAndSelect(
                 'credentialCollectionEntity.credentialCollectionUser',
-                'allCredentialCollectionUsers',
+                'credentialCollectionUserEntity',
+                `
+                    credentialCollectionEntity.id = :id AND
+                    credentialCollectionEntity.tenantId = :tenantId AND
+                    credentialCollectionUserEntity.user.id = :userId AND
+                    credentialCollectionUserEntity.privilegeType = :privilegeType
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: cannot you use ${id}, ${user.tenantId} etc directly? it's template string after all

Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: and  this object below would be redunant if its possible

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection-user/credential-collection-user.entity.ts [hunk 7]
```diff
@@ -21,14 +21,14 @@ export class CredentialCollectionUser {
     @PrimaryGeneratedColumn('uuid')
     id: string;
 
-    @ManyToOne(() => CredentialCollection, (collection) => collection, { onDelete: 'CASCADE' })
+    @ManyToOne(() => CredentialCollection, (collection) => collection.credentialCollectionUser, { onDelete: 'CASCADE' })
```
Comment (socjalis) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection-user/credential-collection-user.entity.ts:24]: Two relations pointing to the same column?

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.controller.ts [hunk 8]
```diff
@@ -85,9 +87,12 @@ export class CredentialCollectionController {
     }
 
     @Delete(':id')
-    delete(@Param('id') id: string) {
+    @HttpCode(HttpStatus.NO_CONTENT)
```
Comment (socjalis) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.controller.ts:90]: is it convention in the project?

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 9]
```diff
@@ -81,56 +82,59 @@ export class CredentialCollectionService {
...
-        }
-
-        return collections;
+    findAllAccessible(user: IUser) {
+        return this.credentialCollectionRepository
+            .createQueryBuilder('credentialCollectionEntity')
+            .innerJoinAndSelect('credentialCollectionEntity.createdBy', 'createdBy')
+            .innerJoinAndSelect(
+                'credentialCollectionEntity.credentialCollectionUser',
+                'credentialCollectionUser',
```
Comment (socjalis) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts:106]: please use where statement with safe use of arguments

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 10]
```diff
@@ -81,56 +82,59 @@ export class CredentialCollectionService {
...
-            relations,
-        });
+        const collection = await this.credentialCollectionRepository
+            .createQueryBuilder('credentialCollectionEntity')
+            .innerJoinAndSelect('credentialCollectionEntity.createdBy', 'createdBy')
+            .innerJoinAndSelect(
+                'credentialCollectionEntity.credentialCollectionUser',
+                'credentialCollectionUser',
+                `
+                    credentialCollectionEntity.id = '${id}' AND
```
Comment (socjalis) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: same thing

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 11]
```diff
@@ -174,29 +178,39 @@ export class CredentialCollectionService {
...
 
-        await collectionUserArray.forEach((item) => {
-            this.credentialCollectionUserRepository.remove(item);
-        });
+            await Promise
+                .all(credentialCollectionUserArray.map((ccu =>
+                    this.credentialCollectionUserRepository.remove(ccu)
+                )))
+                .catch(error => {
+                    throw new InternalServerErrorException(error);
```
Comment (socjalis) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts]: don't throw internal server error exception, errors are handled like this by default already

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts [hunk 12]
```diff
@@ -213,10 +227,36 @@ export class CredentialCollectionService {
...
+                    tenantId: user.tenantId,
+                    credentialCollectionUser: {
+                        userId: user.id,
+                        privilegeType: PrivilegeType.WRITE,
+                    },
+                },
+            });
+
+        if (!collection) {
+            throw new NotFoundException('Could not find credential collection with id: ' + id);
```
Comment (socjalis) [inline runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts:255]: unnecessary message, error code is enough

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection-user/credential-collection-user.entity.ts (hunk 7 above)
Comment (marcin-depcik) [reply runbotics/runbotics-scheduler/src/scheduler-database/credential-collection-user/credential-collection-user.entity.ts:24]: One relation points to CredentialCollection entity and the other points to UserEntity

File: runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.controller.ts (hunk 8 above)
Comment (marcin-depcik) [reply runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.controller.ts:90]: I think so, several resources in `runbotics-scheduler` already use this convention

**Pull request merged**